src/librustc/mir/repr.rs

```diff
@@ -754,35 +743,8 @@ impl<'tcx> Debug for Lvalue<'tcx> {
 ///////////////////////////////////////////////////////////////////////////
 // Scopes
 
-impl Index<ScopeId> for Vec<ScopeData> {
-    type Output = ScopeData;
-
-    #[inline]
-    fn index(&self, index: ScopeId) -> &ScopeData {
-        &self[index.index()]
-    }
-}
-
-impl IndexMut<ScopeId> for Vec<ScopeData> {
-    #[inline]
-    fn index_mut(&mut self, index: ScopeId) -> &mut ScopeData {
-        &mut self[index.index()]
-    }
-}
-
-#[derive(Copy, Clone, Debug, Hash, PartialEq, Eq, RustcEncodable, RustcDecodable)]
-pub struct ScopeId(u32);
-
-impl ScopeId {
-    pub fn new(index: usize) -> ScopeId {
-        assert!(index < (u32::MAX as usize));
-        ScopeId(index as u32)
-    }
-
-    pub fn index(self) -> usize {
-        self.0 as usize
-    }
-}
+newtype_index!(ScopeId, "scope");
+pub const TOPMOST_SCOPE : ScopeId = ScopeId(0);
```

Comment:
Ah yeah I meant to suggest another name in any case (or did I?)



On Wed, Jun 08, 2016 at 12:32:24PM -0700, Eduard-Mihai Burtescu wrote:
> > -
> > -#[derive(Copy, Clone, Debug, Hash, PartialEq, Eq, RustcEncodable, RustcDecodable)]
> > -pub struct ScopeId(u32);
> > -
> > -impl ScopeId {
> > -    pub fn new(index: usize) -> ScopeId {
> > -        assert!(index < (u32::MAX as usize));
> > -        ScopeId(index as u32)
> > -    }
> > -
> > -    pub fn index(self) -> usize {
> > -        self.0 as usize
> > -    }
> > -}
> > +newtype_index!(ScopeId, "scope");
> > +pub const TOPMOST_SCOPE : ScopeId = ScopeId(0);
> 
> This conflicts with #33989 where it's called `ARGUMENT_VISIBILITY_SCOPE`.
> 
> ---
> You are receiving this because you were assigned.
> Reply to this email directly or view it on GitHub:
> https://github.com/rust-lang/rust/pull/34149/files/b1db8a4b44a4427980d9827bb250118f63df9890..a47f1790168c7a920f2c02b39c50e84a80170d6a#r66323400